Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,190 @@
+# Objective 
+The objective of this document is to provide details on how to build a generic queue transport inside CoreWCF so that it can enable various queue binding. This document is based on the discussion outlined here https://github.com/CoreWCF/CoreWCF/issues/164 .
+
+# Why are we building
+When WCF was build there were not many queing technique and there was no concept of cloud. With multiple cloud providers supporting queing technique and many queing technique widely available, we wanted to build a generic transport, so that various queuing mechanism can be supported in CoreWCF.
+
+# Class Diagram
+![CoreWCF Generic Queue](/Documentation/DesignDocs/generic_queue_class_diagram.png?raw=true)
+
+# Class Details
+#### IQueueMiddlewareBuilder and QueueMiddlewareBuilder
+- IQueueMiddlewareBuilder provides signatures for 2 important methods, ```Use``` and ```Build```. 
+```csharp
+internal interface IQueueMiddlewareBuilder 
+{ 
+   IServiceProvider QueueServices { get; set; } 
+   IDictionary<string, object> Properties { get; } 
+   IQueueMiddlewareBuilder Use(Func<QueueMessageDispatcherDelegate, QueueMessageDispatcherDelegate> middleware); 
+   QueueMessageDispatcherDelegate Build(); 
+} 
+```
+- ```Use``` adds middleware delegates to the Queue pipeline 
+- ```Build``` builds the delegate used by this application to process requests received from the queue.
+#### QueueMiddleWareBuilderExtension
+- QueueMiddleWareBuilderExtension defines ```UseMiddleware``` method which specifically search for ```InvokeAsync``` method in each middleware class and construct it as delegate to be invoked by the middleware pipeline. This is similar to the NetTcp class UseMiddlewareFramingConnectionHandshakeExtensions and the ASP.NET Core UseMiddlewareExtensions..
```

Comment:
I just noticed you use triple backticks when inserting inline code fragments. Inline you use a single back tick \`like this\` to produce `like this`.  Can you fix throughout the document. I just noticed we don't define `QueueMessageDispatcherDelegate` anywhere. You should probably add it's definition in the code snippet which contains the `IQueueMiddlewareBuilder` delegate.

```suggestion
The `QueueMiddleWareBuilderExtension` class defines the `UseMiddleware` method which searches the provided `TMiddleware` generic parameter type for an `InvokeAsync` method with a method signature that matches the `QueueMessageDispatcherDelegate` delegate. It expects a constructor to exist which takes as the first parameter an instance of a `QueueMessageDispatcherDelegate` delegate, and any remaining parameters to match the optional additional args. The `QueueMessageDispatcherDelegate` delegate passed to the constructor is the next middleware to call when the middleware doesn't terminate the dispatch chain. This is similar to the NetTcp class `UseMiddlewareFramingConnectionHandshakeExtensions`, and the ASP.NET Core `UseMiddlewareExtensions` class.
```